tidy3d/plugins/design/method.py

```diff
@@ -175,28 +110,466 @@ class MethodGrid(MethodIndependent):
+        utility = UtilityFunction(kind=self.acq_func, kappa=self.kappa, xi=self.xi)
+        opt = BayesianOptimization(
+            f=run_fn, pbounds=boundary_dict, random_state=self.rng_seed, allow_duplicate_points=True
+        )
+
+        # Run variables
+        arg_list = []
+        total_aux_out = []
+        result = []
+        for _ in range(self.initial_iter):
+            next_point = opt.suggest(utility)
+            self._force_int(next_point, parameters)
+            self._handle_param_convert(param_converter, [next_point])
+            arg_list.append(next_point)
+
+        init_output, aux_out = self._extract_output(run_fn(arg_list))
+        self._flatten_and_append(aux_out, total_aux_out)
+
+        for next_point, next_out in zip(arg_list, init_output):
+            result.append(next_out)
+            self._handle_param_convert(invert_param_converter, [next_point])
+            opt.register(params=next_point, target=next_out)
+
+        # Handle subsequent iterations sequentially as BayOpt package does not allow for batched non-random predictions
+        for _ in range(self.n_iter):
+            next_point = opt.suggest(utility)
+            self._force_int(next_point, parameters)
+            self._handle_param_convert(param_converter, [next_point])
+            next_out, aux_out = self._extract_output(run_fn([next_point]))
+            result.append(next_out[0])
+            self._flatten_and_append(aux_out, total_aux_out)
+            self._handle_param_convert(invert_param_converter, [next_point])
+            opt.register(params=next_point, target=next_out[0])
+
+        if console is not None:
+            console.log(
+                f"Best Result: {opt.max['target']}\n"
+                f"Best Parameters: {' '.join([f'{param}: {val}' for param, val in opt.max['params'].items()])}\n"
+            )
+
+        # Output fn_args from the BO.opt object - getting results in situ as opt changes type to float
+        fn_args = []
+        for output in opt.res:
+            fn_args.append(output["params"])
+
+        return fn_args, result, total_aux_out, opt
+
+
+class MethodGenAlg(MethodOptimize, ABC):
+    """A standard method for performing genetic algorithm search"""
+
+    # Args for the user
+    solutions_per_pop: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of solutions per population.",
+        description="The number of solutions to be generated for each population.",
+    )
+
+    n_generations: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of generations.",
+        description="The maximum number of generations to run the genetic algorithm.",
+    )
+
+    n_parents_mating: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of parents mating.",
+        description="The number of solutions to be selected as parents for the next generation.",
+    )
+
+    stop_criteria: Optional[pd.constr(regex=r"\b(?:reach|saturate)_\d+\b")] = pd.Field(
```

Comment:
I'd probably prefer we just define a stop_critera_type as a literal "reach" or "saturate" and then the stop_critera number as an int / separate field

Reply:
reason is I'm not sure how this regex will work with our docs, it seems a bit tricky.

Reply:
Fair point, changed to two fields and added a check that the fields combine successfully.